### PR TITLE
support for multiple SMF profiles

### PR DESCRIPTION
**Pull request by UpadhyayaAnkur**

### CODE REVIEW

File: consumer/sm_context.go [hunk 1]
```diff
@@ -24,16 +25,67 @@ import (
 	"github.com/omec-project/openapi/models"
 )
 
+func getServingSmfIndex(smfNum int) (servingSmfIndex int) {
+	servingSmfIndexStr := os.Getenv("SERVING_SMF_INDEX")
```
Comment (thakurajayL) [inline consumer/sm_context.go:29]: Why we are using ENV variable for this ? Each AMF instance can have its own global variable which can do round robin on the list of SMFs.

File: consumer/sm_context.go [hunk 2]
```diff
@@ -24,16 +25,67 @@ import (
...
+	//iterate over nf instances to ignore failed NF
+	for _, inst := range smCtxt.SmfProfiles {
+		if inst.NfInstanceId != ignoreSmfId {
+			altSmfInst = append(altSmfInst, inst)
+		}
+	}
+
+	if len(altSmfInst) > 0 {
+		smCtxt.SmfProfiles = altSmfInst
+		nfProfile := altSmfInst[0]
```
Comment (thakurajayL) [inline consumer/sm_context.go:49]: - IF we always take 0th NF profile then why we create all alternate SMF list ?

File: consumer/sm_context.go [hunk 3]
```diff
@@ -24,16 +25,67 @@ import (
...
+	nrfUri := ue.ServingAMF().NrfUri
+	param := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{
+		ServiceNames: optional.NewInterface([]models.ServiceName{models.ServiceName_NSMF_PDUSESSION}),
+		Dnn:          optional.NewString(smCtxt.Dnn()),
+		Snssais:      optional.NewInterface(util.MarshToJsonString([]models.Snssai{smCtxt.Snssai()})),
+	}
+
+	result, err := SendSearchNFInstances(nrfUri, models.NfType_SMF, models.NfType_AMF, &param)
+	if err != nil {
+		return nil
```
Comment (thakurajayL) [inline consumer/sm_context.go:69]: Log the error.

File: context/sm_context.go [hunk 4]
```diff
@@ -39,6 +39,8 @@ type SmContext struct {
 	// for duplicate pdu session id handling
 	UlNASTransportVal *nasMessage.ULNASTransport
 	DuplicatedVal     bool
+
+	SmfProfiles []models.NfProfile
```
Comment (thakurajayL) [inline context/sm_context.go:43]: Temp solution. We need to use NRF cache lib code.

File: consumer/sm_context.go (hunk 2 above)
Comment (UpadhyayaAnkur) [reply consumer/sm_context.go:49]: In the code above, we maintain only good SMF instances in amf context and remove non-reachable SMF instances(ignored one). So, it is safe to take 1st(0th index) element from remaining alternate SMF instances

File: consumer/sm_context.go (hunk 3 above)
Comment (UpadhyayaAnkur) [reply consumer/sm_context.go:69]: This code not used as of now and shall be later employed for NRF lib. So, bound to change.

File: context/sm_context.go (hunk 4 above)
Comment (UpadhyayaAnkur) [reply context/sm_context.go:43]: Yes, this is temporary fix until NRF cache lib codes is not available

File: consumer/sm_context.go (hunk 1 above)
Comment (UpadhyayaAnkur) [reply consumer/sm_context.go:29]: This is temporary solution, with global SMF list via NRF cache changes, this code shall be removed.

**Comment by UpadhyayaAnkur** [comment]
retest this please

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**